Added rust bindings

sp1ritCS opened the pull request.

sp1ritCS committed:
bindings: added rust bindings

sp1ritCS committed:
Merge branch 'main' into rust

sp1ritCS committed:
examples: rust: fixed crate path & unprintable Option<GString>

Rafostar commented on bindings/meson.build:
```diff
@@ -0,0 +1,9 @@
+build_rust = not get_option('rs').disabled()
+if build_rust
+	if not build_gir
+		error('cannot build "rust" without "introspection"')
```
rs auto + gir auto/disabled = error ?

Should probably find programs `cargo` and `generator.py` here, have `build_rust` set depending on their found() values and error when enabled only. Please see how this is done for `vapi` in project.

sp1ritCS replied:
should be fixed. I naively assumed that meson "features" just behave like booleans with fancy name

Rafostar commented on bindings/rust/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "gtuber"
+version = "0.0.1"
```
Please do not hardcode version. Can this file be turned into template?

Rafostar commented on bindings/rust/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+version = "0.0.1"
+edition = "2018"
+description = "Rust bindings for gtuber"
+license = "LGPL-3.0-or-later"
```
2.1?

sp1ritCS replied:
I just remember talking about lgpl for gtuber in #clapper. Didn't remember the version, so I just changed the A from my template to L :D

should be fixed now

Rafostar commented on bindings/meson.build:
```diff
@@ -0,0 +1,9 @@
+	if not build_gir
+		error('cannot build "rust" without "introspection"')
+	else
+		subdir('rust')
```
Can I ask for consistent 2 spaces indentation in all `meson.build` files, please?

Rafostar commented on meson.build:
```diff
@@ -1,6 +1,6 @@
 project('gtuber', 'c',
   version: '0.0.1',
-  meson_version: '>=0.53.0',
+  meson_version: '>=0.58.0',
```
What required a version bump? This is a question only.

sp1ritCS replied:
meson.global_source_root() & str.replace()

Rafostar commented on bindings/rust/Gir.toml:
```diff
@@ -0,0 +1,26 @@
+version = "0.0"
+work_mode = "normal"
+
+girs_directories = ["/usr/share/gir-1.0/"]
```
Can this be also templated and set from meson install gir-1.0 dir accordingly?

Rafostar commented on bindings/rust/meson.build:
```diff
@@ -0,0 +1,37 @@
+
+generator = find_program('gir/generator.py')
+run_command(generator,
+	'--gir-files-directories', '/usr/share/gir-1.0/', gtuber_build_dir,
```
hardcoded path.

Rafostar commented on bindings/rust/sys/Gir.toml:
```diff
@@ -0,0 +1,11 @@
+version = "0.0"
+work_mode = "sys"
+
+girs_directories = ["/usr/share/gir-1.0/"]
```
hardcoded path.

Rafostar commented on bindings/rust/meson.build:
```diff
@@ -0,0 +1,37 @@
+	current_srcdir
+)
+
+crate_dir = join_paths(datadir, 'cargo/registry/gtuber-0.0.1/')
```
hardcoded version.

Rafostar commented on bindings/rust/sys/Gir.toml:
```diff
@@ -0,0 +1,11 @@
+[options]
+library = "Gtuber"
+version = "0.0"
```
hardcoded version.

sp1ritCS committed:
Minor fixes requested by @Rafostar

- downgraded license to LGPL-2.1
- fixed wierd feature behavior (I just asumed they behave like booleans)
- added /usr/local/ prefix as addtional gir path

sp1ritCS committed:
removed nullability from StreamExt::url()

sp1ritCS committed:
actions: Added deploy action that automatically submits rust bindings to sp1ritCS/gtuber-rs

Rafostar commented:
Under agreement of both parties, gtuber rust bindings are available in [https://github.com/sp1ritCS/gtuber-rs](https://github.com/sp1ritCS/gtuber-rs), since closing. Many thanks for contributing and making those Rust bindings possible!
